---
title: "Altruistic Behavior in Animal Populations"
output:
  html_document:
    keep_md: false
---

```{r include = FALSE}
library(simaltruist)
knitr::opts_chunk$set(out.width = "90%", fig.align = "center")
```

<br><br>
The R package **simaltruist** provides a convenient interfacing for running simulations to explore the spread of genes for altruism in an animal population.

## Scenario

Coming soon!

## Installation

Install the package from Git Hub:

```{r eval = FALSE}
devtools::install_github("homerhanumat/simaltruist")
```

Then attach the package:

```{r eval = FALSE}
library(simaltruist)
```

## Basic Usage

Here is a simple example:

```{r cache = TRUE, fig.cap = "A graph for free!"}
pop <- simulate(sim_gens = 400,
                warner_death_prob = 0.8)
```

Here's the data stored in `pop`:

```{r echo = FALSE}
DT::datatable(pop)
```

<br><br>

## Writing and Using Custom Functions

The `simulate()` function provides for a certain amount of customization of two important phases in the simulation of a single generation:

* the way in which mating couples are formed;
* how warners behave when the population is under attack by a predator.

Both of these behaviors may be controlled and programmed by the user, by means of the parameters `mating_behavior` and `attack_behavior`.

### Mating Behavior

By default, the behavior of a mating couple is determined by the function `randomMatches()`, the source code of which is given below:

```{r eval = FALSE}
randomMatches <- function(individuals, number_of_couples) {
  females <- subset(individuals, sex == "F")
  breedingFemales <- females[sample(
    1:nrow(females),
    size = number_of_couples,
    replace = FALSE), ]
  males <- subset(individuals,sex == "M")
  chosenMales <- males[sample(
    1:nrow(males),
    size = nrow(breedingFemales),
    replace = TRUE), ]
  list(females = breedingFemales, males = chosenMales)
}
```

As the code shows, the default behavior is for fertile females to select randomly from the set of males.

As for the arguments, `individuals` is a data frame that is computed at the beginning of each generation.  Each row represents a living member of the population.  The columns are:

* `id`:  a string giving the ID-number of the individual;
* `sex`:  the sex of the individual;
* `warner`:  a numerical variable with values 0, 1, or 2, giving the number of altruistic alleles the individual possesses.
* `mom`: a string giving the ID-number of the individual's mother;
* `dad`: a string giving the ID-number of the individual's mother.

It is possible to write one's own mating function and pass it to `simulate()` via the `mating_behavior` parameter.  An example of such a custom function is exported by **simaltruist** as `sexualSelection()`, the source code for which maybe be viewed [here](https://github.com/homerhanumat/simaltruist/blob/master/R/reproduce.R).

The formal parameters of the function are shown in the definition below:

```{r eval = FALSE}
sexualSelection <- function(
  individuals,
  number_of_couples,
  prefs = matrix(1, nrow = 3, ncol = 3)
  ) {
  ## function body
}
```


Examples of its use are provided in the package documentation, e.g.:

```{r fig.cap="Sexual selection:  opposites attract!"}
pop <- simulate(sim_gens = 200,
                mating_behavior = list(
                  fn = sexualSelection,
                  args = list(
                    prefs = matrix(
                      c(1, 5, 10, 1, 1, 1, 10, 5, 1),
                      nrow = 3,
                      ncol = 3))))
```

In the above example, the preference matrix is:

                male0     male1     male2
-----           ------    ------    --------
*female0*       1         5         10
*female1*       1         1         1
*female02*      10        5         1

We see that 0-allele females prefer 1-allele and 2-allele males.  In fact when presented with three males, one of each genotype, she is five times as likely to prefer the 1-allele male over the 0-allele male, and ten times as likely to prefer the 2-allele male over the 0-allele male. Zero-allele females have a reverse preference, and single-allele females are indifferent in their choice.

Note that `args` specifies only the value of the `prefs` parameter of `sexualSelection`.  The other two parameters of this function---`individuals` and `number_of_couples` are computed by `simulate()` and provided to `sexualSelection()` when it is called.

For custom mating functions the possible provided parameters are:

* `individuals`, the data frame described above;
* `number_of_couples`
* `relMatrix`, a matrix (described below) of relationship-degree between living individuals.

Note that a custom mating function must return a named list containing two elements:

* `females`:  a character vector consisting of the id numbers of female mates;
* `males`:  a character vector consisting of the id numbers of male mates.

Corresponding elements of the two vectors represent a single couple.

### Attack Behavior

By default the behavior of a warner when the population is under attack is governed by the function `warnRelatives()`, described in the package documentation.  Again it is possible for the user to pass in a modified version of this function or an entirely new function, via the parameter `attack_behavior`.

An example of an alternative attack-behavior function is the exported function [`warnWarners()`](https://github.com/homerhanumat/simaltruist/blob/master/R/attack.R), in which each warner warns only other warners, without regard to their relationship to him or her:

```{r cache = TRUE, fig,cap = "Here, each 2-allele warner can warn up to seven warners."}
pop <- simulate(sim_gens = 100,
                attack_behavior = list(
                  fn = warnWarners,
                  args = list(
                    number_warned = 7,
                    dominant = TRUE)))
```


Again it is only necessary to set the values of parameters that are not automatically provided by the `simulate()` function.  Provided values are as follows:
  
* `individuals`, described previously;
* three parameters provided to `simulate()`:
    * `warner_death_prob`
    * `nonwarner_death_prob`
    * `hider_death_prob`
* `relMatrix`, a matrix with row and columns names corresponding to the ID-numbers of all living individuals.  The cells give the degree of relationship between any given pair of individuals.

Note that a custom attack function must return a numerical vector of death probabilities, with the $i^{\text{th}}$ element corresponding to row $i$ of `individuals`.
    



