---
title: "Altruistic Behavior in Animal Populations"
output:
  html_document:
    keep_md: false
---

```{r include = FALSE}
library(simaltruist)
knitr::opts_chunk$set(out.width = "90%", fig.align = "center")
```

<br><br>
The R package **simaltruist** provides a convenient interface for running simulations to explore the spread of genes for altruism in an animal population.

## Scenario

Coming soon!

## Installation

Install the package from Git Hub:

```{r eval = FALSE}
devtools::install_github("homerhanumat/simaltruist")
```

Then attach the package:

```{r eval = FALSE}
library(simaltruist)
```

## Basic Usage

Here is a simple example:

```{r cache = TRUE, fig.cap = "A graph for free!"}
pop <- simulate(sim_gens = 400,
                warner_death_prob = 0.8)
```

Here's the data stored in `pop`:

```{r echo = FALSE}
DT::datatable(pop)
```

<br><br>

## Writing and Using Custom Functions

The `simulate()` function provides for a extensive customization of three important phases in the simulation of a single generation:

* the way in which mating couples are formed;
* the chance of death (by cause other than predator attack) in a "culling";
* how warners behave when the population is under attack by a predator.

Both of these behaviors may be controlled and programmed by the user, by means of the parameters `mating_behavior`, `culling_behavior` and `attack_behavior`.

All custom functions have the following values provided by the `simulate()` program, provided that appear as named parameters in definition of the function:

* `individuals`:  a data frame that is computed at the beginning of each generation.  Each row represents a living member of the population.  The columns are:
    * `id`:  a string giving the ID-number of the individual;
    * `sex`:  the sex of the individual;
    * `warner`:  a numerical variable with values 0, 1, or 2, giving the number of altruistic alleles the individual possesses.
    * `mom`: a string giving the ID-number of the individual's mother;
    * `dad`: a string giving the ID-number of the individual's mother.
* `population`: a data frame on the population at each generation up to the current one.  Columns are:
    * `populationSize`;
    * `males`:  total number of males in the current generation;
    * `males0`:  number of males with 0 altruist alleles;
    * `males1`:  number of males with 1 altruist allele;
    * `males2`:  number of males with 2 altruist alleles;
    * `females`:  total number of females in the current generation;
    * `females0`:  number of females with 0 altruist alleles;
    * `females1`:  number of females with 1 altruist allele;
    * `females2`:  number of females with 2 altruist alleles
* `relMatrix`: a matrix with row and columns names corresponding to the ID-numbers of all living individuals.  The cells give the degree of relationship between any given pair of individuals.
* `current_gen`:  the current generation (starts with 1);
* all of the following values already passed into the `simulate()` function:
    * `average_litter_size`
    * `birth_rate_natural`
    * `death_rate_natural`
    * `prob_attack`
    * `warner_death_prob`
    * `nonwarner_death_prob`
    * `hider_death_prob`
    * `capacity`

### Mating Behavior

By default, the behavior of a mating couple is determined by the function `randomMatches()`, the source code of which is given below:

```{r eval = FALSE}
randomMatches <- function(individuals, number_of_couples) {
  females <- subset(individuals, sex == "F")
  breedingFemales <- females[sample(
    1:nrow(females),
    size = number_of_couples,
    replace = FALSE), ]
  males <- subset(individuals,sex == "M")
  chosenMales <- males[sample(
    1:nrow(males),
    size = nrow(breedingFemales),
    replace = TRUE), ]
  list(females = breedingFemales, males = chosenMales)
}
```

As the code shows, the default behavior is for fertile females to select randomly from the set of males.

It is possible to write one's own mating function and pass it to `simulate()` via the `mating_behavior` parameter.  Such a function has access to all of the values listed above, as well as `number_of_couples`, the number of couples that should form in the current generation in order to satisfy the required birth rate.

An example of such a custom function is exported by **simaltruist** as `sexualSelection()`, the source code for which maybe be viewed [here](https://github.com/homerhanumat/simaltruist/blob/master/R/reproduce.R).

The formal parameters of the function are shown in the definition below:

```{r eval = FALSE}
sexualSelection <- function(
  individuals,
  number_of_couples,
  prefs = matrix(1, nrow = 3, ncol = 3)
  ) {
  ## function body
}
```


Examples of its use are provided in the package documentation, e.g.:

```{r fig.cap="Sexual selection:  opposites attract!"}
pop <- simulate(sim_gens = 200,
                mating_behavior = list(
                  fn = sexualSelection,
                  args = list(
                    prefs = matrix(
                      c(1, 5, 10, 1, 1, 1, 10, 5, 1),
                      nrow = 3, ncol = 3,
                      byrow = TRUE))))
```

In the above example, the preference matrix is:

                male0     male1     male2
-----           ------    ------    --------
*female0*       1         5         10
*female1*       1         1         1
*female2*       10        5         1

We see that 0-allele females prefer 1-allele and 2-allele males.  In fact when presented with three males, one of each genotype, she is five times as likely to prefer the 1-allele male over the 0-allele male, and ten times as likely to prefer the 2-allele male over the 0-allele male. Zero-allele females have a reverse preference, and single-allele females are indifferent in their choice.

Note that `args` specifies only the value of the `prefs` parameter of `sexualSelection`.  The other two parameters of this function---`individuals` and `number_of_couples` are computed by `simulate()` and provided to `sexualSelection()` when it is called.

Note also that a custom mating function must return a named list containing two elements:

* `females`:  a character vector consisting of the id numbers of female mates;
* `males`:  a character vector consisting of the id numbers of male mates.

Corresponding elements of the two vectors represent a single couple.

### Culling Behavior

In any generation there is a culling phase that represents death by all causes other than a mass attack on the colony by predators.  The death-rate for this phase is computed as a linear function of population-size that is equal to `death_rate_natrual` when the population is zero and is equal to `birth_rate_natural` when the population is at capacity.

The default culling function assigns this computed death-rate `dr` as the probability of death for each individual in the population during the culling phase.

Custom culling functions may be written.  As with custom mating functions have access to all values in the list above, but they also have access to `dr`.  An example of a custom culling function is as follows:

```{r}
weakWarners <- function(individuals, dr) {
  warner <- individuals$warner
  deathProb <- numeric(length(warner))
  deathProb[warner == 0] <- dr
  deathProb[warner == 1] <- dr * 2
  deathProb[warner == 2] <- dr * 5
  deathProb
}
```

This function gives warners a much higher chance of death during the culling phase.  (Perhaps they expend a lot of energy patrolling for predators.)

An example of use of this function is as follows.  It assumes a scenario in which there are no predator attacks.

```{r cache = TRUE, fig.cap = "Weak warners go extinct!"}
pop <- simulate(sim_gens = 200, prob_attack = 0,
                culling_behavior = list(fn = weakWarners,
                                        args = list()))
```

Note that `weakWarners()` has no arguments beyond those whose values are provided by the `simulate()` function, so we set `args` to the empty list.


### Attack Behavior

By default the behavior of a warner when the population is under attack is governed by the function `warnRelatives()`, described in the package documentation.  Again it is possible for the user to pass in a modified version of this function or an entirely new function, via the parameter `attack_behavior`.  A custom attack function has access to any values in the list above, provided that the value appears as a parameter in the definition of the function.

An example of an alternative attack-behavior function is the exported function [`warnWarners()`](https://github.com/homerhanumat/simaltruist/blob/master/R/attack.R), in which each warner warns only other warners, without regard to their relationship to him or her:

```{r cache = TRUE, fig.cap = "Here, each 2-allele warner can warn up to seven warners."}
pop <- simulate(sim_gens = 100,
                attack_behavior = list(
                  fn = warnWarners,
                  args = list(
                    number_warned = 7,
                    dominant = TRUE)))
```


Again it is only necessary to set the values of parameters that are not automatically provided by the `simulate()` function.

Note that a custom attack function must return a numerical vector of death probabilities, with the $i^{\text{th}}$ element corresponding to row $i$ of `individuals`.

## Representation of Relationship

By default the `simulate()` function creates and updates `relMatrix`, a matrix with rows and column names identical to the Ids of living individuals.  Thus---for example---both `relMatrix["123","17"]` and `relMatrix["123","17"]` would give the degree of relationship between individual 123 and individual 17.  When an individual is born it is introduced to the matrix.  Its degree of relationship with itself is set to 1, and the degree of relationship to any sibling is set to 0.5.  The degree of relationship to any other individual $I$ is:

$$\frac{\texttt{relMatrix[mother, I]} + \texttt{relMatrix[mother, I]}}{2}.$$

If the parameter `relationship_method` is set to `"graph"` then the `simulate()` function creates and updates `relGraph`, a family tree for the population in the form of a directed graph with edges from individuals to each of their children.  The vertices are numbered according to the Id of each individual.  Unlike `relMatrix`, `relGraph` keeps on growing, since individuals aren't removed when they die.  The default behavior of warners under attack is governed by `warnRelativesGraph()`, in which warners tip off up to ten relatives, provided that the relative is also less than two edges "distant"  from the warner on the relatioship graph.  To ilustrate, in the abence of interbreeding:

* the distance from an individual to its parent is 1;
* to a half-sibling is 2;
* to a grandparent is 2;
* to an aunt or uncle is 3;
* to a cousin is 4.

`relGraph` replaces `relMatrix` in the set of provided variables for custom functions.

When one's custom functions do not make reference to either `relMatrix` or `relGraph`, it is recommended to set `relationship_method` to `"none"` in order to speed up the simulation.
    



